### ISSUE TITLE
Purge operations result in broken dlist files

### ISSUE THREAD

**Issue by warwickmm** [duplicati/duplicati#3924]
**Duplicati version**: 2.0.4.29

## Description

A purge operation results in broken dlist files, as well as possibly other issues.

## Steps to reproduce

1. Run a backup.
2. Purge a file.  This _appears_ to run without error.  However, notice that most of the dlist files are 541 bytes in size.
3. Recreate the database.  This should result in an exception and reveal that the purge did not behave properly:
```
Failed to process file: duplicati-20190925T020411Z.dlist.zip.aes => Value cannot be null.
Parameter name: stream
Failed to process file: duplicati-20190925T020414Z.dlist.zip.aes => Value cannot be null.
Parameter name: stream
  Downloading file (1.39 KB) ...
```

This appears to have been introduced in pull request #3836, possibly between revisions e8f3bccef8da469f3b00f228a77b1965b16d0781 and 153a97285315d9562f21384365fc45b783d862ec.

See forum discussion [here](https://forum.duplicati.com/t/purge-resulted-in-damaged-dlist-files/8114).

**Comment by warwickmm** [comment]
I can get my tests to pass if I add
```
vol.AddFilelistFile();
```
to the `PurgeFileHander.DoRun` method [here](https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Main/Operation/PurgeFilesHandler.cs#L165).  However, it would be more satisfying if we didn't have to remember to invoke this everywhere.

@BlueBlock, any suggestions here?

**Comment by warwickmm** [comment]
Revision e0899a2a91aab8e982e9674461c0504907ef5845 appears to have been the one that made creation of the filelist non-mandatory.  However, the commit message mentions that it fixes a different issue regarding the "control-files" that can be optionally added.

**Comment by ts678** [comment]
2.0.4.29 has the same problem if using repair to re-upload a missing dlist file. Possibly similar cause?

> C:\>"C:\ProgramData\Duplicati\duplicati-2.0.4.29_canary_2019-09-17\Duplicati.CommandLine.exe" repair "file://C:\ProgramData\Duplicati\duplicati-2.0.4.29_canary_2019-09-17\RUN\purge test 1\\" --dry-run=true --backup-name="purge test 1" --dbpath="C:\ProgramData\Duplicati\duplicati-2.0.4.29_canary_2019-09-17\RUN\SKJBADDCGJ.sqlite" --encryption-module= --compression-module=zip --dblock-size=50mb --no-encryption=true --no-auto-compact=true --log-file-log-level=Profiling --profile-all-database-queries=true --log-file="C:\ProgramData\Duplicati\duplicati-2.0.4.29_canary_2019-09-17\RUN\purge test 1.Profiling.log" --disable-module=console-password-input --console-log-level=DryRun
  Listing remote folder ...
[Dryrun]: would re-upload fileset duplicati-20190925T224456Z.dlist.zip, with size 233 bytes, previous size 758 bytes

https://github.com/duplicati/duplicati/blob/c412bc9cae0db9780f7c219bed34c17bef5ef1c8/Duplicati/Library/Main/Operation/RepairHandler.cs#L294-L303

**Comment by warwickmm** [comment]
Yes, my suspicion is that the `FilesetVolumeWriter` class was modified and usages weren't all updated accordingly.

**Comment by ts678** [comment]
My suspicion based on a dnSpy run of --dry-run repair on 2.0.4.30 is that the filelist JSON is written into a memory stream (I can watch 440 bytes of wanted JSON accumulating in charBuffer) which goes nowhere.

Previously the design (which I have not looked over before, so someone else ought to take a look as well) wouldn't have needed to do any special handling to copy the various non-manifest zip contents into dlist because it wrote them there originally via m_compression from the base class, which is where manifest is added. Result (not yet tested with a debugger on old release) is manifest file by itself, and rest got tossed.

[add method to add the filelist file](https://github.com/duplicati/duplicati/commit/e0899a2a91aab8e982e9674461c0504907ef5845#diff-bd7255b05b4f39949e263b1e06d63901) might be be where the filelist was added (but maybe to nowhere). What do @BlueBlock and others think? Does this fit chronology? It does by date, but how to timeline it exactly?

**Comment by ts678** [comment]
[add method to add filelist file  -- This is part of correcting the way the dlist zip file is created. We now have a method to take the memorystream and put the file into dlist.](https://github.com/duplicati/duplicati/commit/3b5af1cf01aa6fbebc9296142ff67d280353b24b) is possibly why backup still works, because a call to AddFilelistFile() was added to Backup/UploadRealFilelist.cs and Backup/UploadSyntheticFilelist.cs, although I'm not sure the latter is correct because its file list copying gets done before CreateFilesetAsync, AppendFilesFromPreviousSetAsync, and WriteFilesetAsync which seem like they'd make a different file list. Breaking this doesn't worry me too much immediately because it's already inaccessible as explained [here](https://github.com/duplicati/duplicati/issues/2506#issuecomment-463622742).

What worries me is that when I search source for strings containing WriteFileset, I find the two above as WriteFilesetAsync, but I also find WriteFileset at Duplicati\Library\Main\Operation\PurgeFilesHandler.cs and Duplicati\Library\Main\Operation\RepairHandler.cs which don't have AddFilelistFile() calls, possibly explaining why purge and repair produce bad dlist files. Maybe their code should be similarly modified?

**Comment by warwickmm** [comment]
I have a potential fix that I'll be running through some tests and will report back if the results are promising.

**Comment by drwtsn32x** [comment]
@warwickmm  - great! I will be happy to test

**Comment by warwickmm** [comment]
Thanks @drwtsn32x.  I'm still running some tests but will create a pull request if things go well.

Also, is anyone having any issues with the purge operation?  I'm seeing that all files _except_ the specified file are removed from the backup sets.  I thought it would work the other way, where the specified file would be removed.  Perhaps I'm doing something wrong in my tests.

**Comment by warwickmm** [comment]
@drwtsn32x, I submitted a candidate fix in pull request #3925.  If you're able to help test, that would be much appreciated!

**Comment by drwtsn32x** [comment]
@warwickmm yep going to give this a shot now - thanks for your efforts!

**Comment by drwtsn32x** [comment]
> Also, is anyone having any issues with the purge operation? I'm seeing that all files except the specified file are removed from the backup sets.

I haven't tried it the way you did... I have been using `--include=<path to purge>`

**Comment by drwtsn32x** [comment]
Testing - so far it looks promising.  The new dlist files appear to be sane size.

Will let you know once my testing is complete!

**Comment by warwickmm** [comment]
Thanks!  Can you leave any feedback on the pull request?  That way it will be easier to follow the conversation and history of changes.

**Comment by warwickmm** [comment]
Fixed in pull request #3925.
